components/esp_modem/src/esp_modem_dce.cpp

```diff
@@ -27,24 +29,50 @@ bool DCE_Mode::set(DTE *dte, ModuleIf *device, Netif &netif, modem_mode m)
     case modem_mode::UNDEF:
         break;
     case modem_mode::COMMAND_MODE:
-        if (mode == modem_mode::COMMAND_MODE) {
-            return false;
-        }
-        netif.stop();
-        if (!device->set_mode(modem_mode::COMMAND_MODE)) {
-            return false;
-        }
-        dte->set_read_cb([&](uint8_t *data, size_t len) -> bool {
-            ESP_LOG_BUFFER_HEXDUMP("esp-modem: debug_data", data, len, ESP_LOG_INFO);
-            return false;
-        });
-        netif.wait_until_ppp_exits();
-        dte->set_read_cb(nullptr);
-        if (!dte->set_mode(modem_mode::COMMAND_MODE)) {
-            return false;
+        {
+            if (mode == modem_mode::COMMAND_MODE) {
+                return false;
+            }
+            if (mode == modem_mode::CMUX_MODE) {
+                netif.stop();
+                netif.wait_until_ppp_exits();
+                if (!dte->set_mode(modem_mode::COMMAND_MODE)) {
+                    return false;
+                }
+                mode = m;
+                return true;
+            }
+            netif.stop();
+            SignalGroup signal;
+            dte->set_read_cb([&](uint8_t *data, size_t len) -> bool {
```

Comment:
I would suggest capture explicitly here, since it's a callback we may have lifetime issues if some references go out of scope. ASIO code uses that lifetime extension idea using `std::shared_from_this` to guarantee lifetime.

Reply:
The implicit capture was needed for logging as the `ESP_LOGx()` variants need statics and sometimes are defined and sometimes not (based on configured log severity).
I think we can capture the `signal` part only, hope it won't be necessary to use `shared_from_this`...

Reply:
captured only `signal`. Please resolve if you agree.

Reply:
> static don't need to be captured.

But `signal` is not static in this context, it's just an automatic variable. Or are you implying that we should make it static ?

Reply:
So I made the automatic variable a `shared_ptr<>` that's passed to the lambda as `weak_ptr<>`. This should address concerns related to local variable lifetime.